src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,621 @@
+      while ( rsTables->next() )
+      {
+        QgsHanaLayerProperty layerProperty;
+        layerProperty.schemaName = QgsHanaUtils::toQString( rsTables->getNString( 1 ) );
+        layerProperty.tableName = QgsHanaUtils::toQString( rsTables->getNString( 2 ) );
+        layerProperty.geometryColName = "";
+        layerProperty.tableComment = QgsHanaUtils::toQString( rsTables->getNString( 3 ) );
+
+        addNewLayer( list, layerProperty, true );
+      }
+      rsTables->close();
+    }
+
+    // Read views
+    sql = QStringLiteral( "SELECT SCHEMA_NAME, VIEW_NAME, COLUMN_NAME, DATA_TYPE_NAME, VIEW_OID, VIEW_COMMENTS FROM "
+                          "(SELECT * FROM SYS.VIEW_COLUMNS WHERE VIEW_OID IN (SELECT OBJECT_OID FROM OWNERSHIP WHERE "
+                          "OBJECT_TYPE = 'VIEW' AND OWNER_NAME LIKE '%1')) "
+                          "INNER JOIN "
+                          "(SELECT VIEW_OID AS VIEW_OID_2, COMMENTS AS VIEW_COMMENTS FROM SYS.VIEWS WHERE "
+                          "IS_VALID = 'TRUE' AND SCHEMA_NAME IN (SELECT SCHEMA_NAME FROM SYS.SCHEMAS WHERE "
+                          "HAS_PRIVILEGES = 'TRUE') AND SCHEMA_NAME NOT LIKE_REGEXPR 'SYS|_SYS.*|UIS|SAP_XS|SAP_REST|HANA_XS') "
+                          "ON VIEW_OID = VIEW_OID_2 AND SCHEMA_NAME LIKE '%2' AND DATA_TYPE_NAME LIKE_REGEXPR '%3'" )
+          .arg(
+            userTablesOnly ? "CURRENT_USER" : "%",
+            schema.isEmpty() ? "%" : schema,
+            allowGeometrylessTables ? "" : "ST_GEOMETRY|ST_POINT" );
+    ResultSetRef rsViews = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+    while ( rsViews->next() )
+    {
+      QgsHanaLayerProperty layerProperty;
+      layerProperty.schemaName = QgsHanaUtils::toQString( rsViews->getNString( 1 ) );
+      layerProperty.tableName = QgsHanaUtils::toQString( rsViews->getNString( 2 ) );
+      layerProperty.geometryColName = ( QgsHanaUtils::toQString( rsViews->getNString( 4 ) ) != "ST_GEOMETRY" ) ? ""
+                                      : QgsHanaUtils::toQString( rsViews->getNString( 3 ) );
+      layerProperty.tableComment = QgsHanaUtils::toQString( rsViews->getNString( 6 ) );
+      layerProperty.isView = true;
+
+      addNewLayer( list, layerProperty, layerProperty.geometryColName.isEmpty() );
+    }
+    rsViews->close();
+
+    int size = list.size();
+    for ( int i = 0; i < size; ++i )
+    {
+      QgsHanaLayerProperty &lp1 = list[i];
+      bool found = false;
+      for ( int j = 0; j < size; ++j )
+      {
+        if ( i != j )
+        {
+          const QgsHanaLayerProperty &lp2 = list.at( j );
+          if ( lp1.schemaName == lp2.schemaName && lp1.tableName == lp2.tableName )
+          {
+            found = true;
+            break;
+          }
+        }
+      }
+
+      if ( !found )
+        lp1.isUnique = true;
+    }
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
```

Comment:
`addNewLayer()` and the uniqueness check are both `O(n^2)`. We could use `QMultiHash<QPair<QString, QString>, QgsHanaLayerProperty>` to make the duplicate/uniqueness checks efficient. In the end we can easily convert that to a vector via `QMultiHash::values()`.

Reply:
Done in 84beb39.